### PR TITLE
Added error code for overlapping function signatures

### LINKED ISSUE #17570: Use a more specific error code for overlapping function signatures
Hi, would it be possible to add a more specific error code for the `Overloaded function signature x will never be matched` error?

I want to be able to disable it in the config without also disabling every other `misc` error.

### PR DESCRIPTION
**Pull request by katconnors**
Closes #17570. This is my first contribution to mypy! 🐍 
Added an error code for overlapping function signatures.
Test in check-errorcodes.test is a derivative of this post: https://stackoverflow.com/q/69341607

### CODE REVIEW

**Comment by katconnors** [comment]
mypy_primer (2) timed out- wonder if it had the same issue as here:
https://github.com/python/mypy/pull/17588#issuecomment-2253537869

File: mypy/errorcodes.py [hunk 1]
```diff
@@ -262,6 +262,12 @@ def __hash__(self) -> int:
     default_enabled=False,
 )
 
+OVERLOADED_FUNCTION_MATCHING: Final[ErrorCode] = ErrorCode(
+    "overloaded-function-matching",
+    "Warn user about signature matching for overloaded functions.",
+    "General",
```
Comment (AlexWaygood) [inline mypy/errorcodes.py:268]: if you do this, it makes it backwards compatible: people who want more specific `type: ignore` comments can use the new error code, but other people's existing `type: ignore` comments won't be broken

```suggestion
    "General",
    sub_code_of=MISC,
```

Comment (katconnors) [reply mypy/errorcodes.py:268]: great, thanks for pointing that out!

File: mypy/checker.py [hunk 2]
```diff
@@ -768,6 +768,7 @@ def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
                     continue
 
                 if overload_can_never_match(sig1, sig2):
+
```
Comment (JelleZijlstra) [inline mypy/checker.py:771]: Remove unnecessary changes (there's another one below).

File: mypy/errorcodes.py [hunk 3]
```diff
@@ -273,6 +274,14 @@ def __hash__(self) -> int:
 # This is a catch-all for remaining uncategorized errors.
 MISC: Final[ErrorCode] = ErrorCode("misc", "Miscellaneous other checks", "General")
 
+OVERLOADED_FUNCTION_MATCHING: Final[ErrorCode] = ErrorCode(
+    "overloaded-function-matching",
```
Comment (JelleZijlstra) [inline mypy/errorcodes.py:278]: This name doesn't seem great, maybe something like `overload-cannot-match`?

File: docs/source/error_code_list.rst [hunk 4]
```diff
@@ -1149,6 +1149,26 @@ types you expect.
 
 See :ref:`overloading <function-overloading>` for more explanation.
 
+
+.. _code-overloaded-function-matching:
+
+Check error code of overload function signature match
```
Comment (JelleZijlstra) [inline docs/source/error_code_list.rst:1155]: The documentation should say something about when this error happens and what to do about it.

**Comment by katconnors** [comment]
@JelleZijlstra made the requested changes, if you don't mind taking another look. Thanks in advance!

File: mypy/message_registry.py [hunk 5]
```diff
@@ -362,3 +362,8 @@ def with_additional_msg(self, info: str) -> ErrorMessage:
 TYPE_ALIAS_WITH_AWAIT_EXPRESSION: Final = ErrorMessage(
     "Await expression cannot be used within a type alias", codes.SYNTAX
 )
+
+# Overloads
+OVERLOADED_FUNCTION_SIGNATURE: Final = ErrorMessage(
```
Comment (JelleZijlstra) [inline mypy/message_registry.py:367]: This seems unused.

File: mypy/errorcodes.py [hunk 6]
```diff
@@ -273,6 +273,14 @@ def __hash__(self) -> int:
 # This is a catch-all for remaining uncategorized errors.
 MISC: Final[ErrorCode] = ErrorCode("misc", "Miscellaneous other checks", "General")
 
+OVERLOAD_CANNOT_MATCH: Final[ErrorCode] = ErrorCode(
+    "overload-cannot-match",
+    "Warn user about signature matching for overloaded functions.",
```
Comment (JelleZijlstra) [inline mypy/errorcodes.py:278]:
```suggestion
    "Warn if an @overload signature can never be matched",
```
Consistent formatting with the next one.

File: docs/source/error_code_list.rst [hunk 7]
```diff
@@ -1149,6 +1149,32 @@ types you expect.
 
 See :ref:`overloading <function-overloading>` for more explanation.
 
+
+.. _code-overload-cannot-match:
+
+Check error code of overload function signature match
```
Comment (JelleZijlstra) [inline docs/source/error_code_list.rst:1155]:
```suggestion
Check for overload signatures that cannot match [overload-cannot-match]
```

File: docs/source/error_code_list.rst [hunk 8]
```diff
@@ -1149,6 +1149,32 @@ types you expect.
...
+--------------------------------------------------------------------------
+
+In the case of an overloaded function, if one of the signatures is never accessible, this error may occur.
+An example where this can occur is with the utilization of floats and int types, due to the behavior of integers matching floats in mypy.
```
Comment (JelleZijlstra) [inline docs/source/error_code_list.rst:1159]: An example involving float/int isn't great because it relies on another subtle part of the type system. Better to give an example that is more basic, such as having the first overload take `object` and the second `int`.

Comment (katconnors) [reply docs/source/error_code_list.rst]: Valid point. I made some edits to this- please let me know if that looks ok.

**Comment by katconnors** [comment]
Gentle reminder for review, @JelleZijlstra or @AlexWaygood. Thanks both!

File: docs/source/error_code_list.rst [hunk 9]
```diff
@@ -1149,6 +1149,32 @@ types you expect.
...
+    from typing import overload, Union
+
+    @overload
+    def process(response1: object,response2: object) -> object:
+        ...
+    @overload
+    def process(response1: int,response2: int) -> int: # E: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader  [overload-cannot-match]
+        ...
+
+    def process(response1,response2)-> Union[object,int]:
```
Comment (JelleZijlstra) [inline docs/source/error_code_list.rst:1175]:
```suggestion
    def process(response1: object, response2: object) -> object:
        ...
    @overload
    def process(response1: int, response2: int) -> int: # E: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader  [overload-cannot-match]
        ...

    def process(response1: object, response2: object) -> object:
```

File: docs/source/error_code_list.rst [hunk 10]
```diff
@@ -1149,6 +1149,32 @@ types you expect.
...
+    return response1 + response2
```
Comment (JelleZijlstra) [inline docs/source/error_code_list.rst:1176]:
```suggestion
        return response1 + response2
```

File: docs/source/error_code_list.rst [hunk 11]
```diff
@@ -1149,6 +1149,32 @@ types you expect.
...
 
+
+.. _code-overload-cannot-match:
+
+Check for overload signatures that cannot match [overload-cannot-match]
+--------------------------------------------------------------------------
+
+In the case of an overloaded function, if one of the signatures is never accessible, this error may occur.
+An example where this can occur is with the utilization of object and int types.
+Consider swapping the declaration of the two types so that the narrower signature is declared before the broader signature.
```
Comment (JelleZijlstra) [inline docs/source/error_code_list.rst:1160]:
```suggestion
Warn if an ``@overload`` variant can never be matched, because an earlier
overload has a wider signature. For example, this can happen if the two
overloads accept the same parameters and each parameter on the first overload
has the same type or a wider type than the corresponding parameter on the second
overload.
```

**Review by JelleZijlstra — APPROVED** [review]

**Comment by katconnors** [comment]
Looks like checks have passed, and this can be merged when someone has time.
